packages/ses/src/module-load.js

```diff
@@ -145,83 +139,72 @@ function* loadWithoutErrorAnnotation(
   const { importHook, importNowHook, moduleMap, moduleMapHook, moduleRecords } =
     weakmapGet(compartmentPrivateFields, compartment);
 
+  if (mapHas(moduleRecords, moduleSpecifier)) {
+    return mapGet(moduleRecords, moduleSpecifier);
+  }
+
   // Follow moduleMap, or moduleMapHook if present.
-  let aliasNamespace = moduleMap[moduleSpecifier];
-  if (aliasNamespace === undefined && moduleMapHook !== undefined) {
-    aliasNamespace = moduleMapHook(moduleSpecifier);
+  let moduleDescriptor = moduleMap[moduleSpecifier];
+  if (moduleDescriptor === undefined && moduleMapHook !== undefined) {
+    moduleDescriptor = moduleMapHook(moduleSpecifier);
+  }
+  if (moduleDescriptor === undefined) {
+    const moduleHook = selectImplementation(importHook, importNowHook);
+    if (moduleHook === undefined) {
+      const moduleHookName = selectImplementation(
+        'importHook',
+        'importNowHook',
+      );
+      Fail`${moduleHookName} needed to load module ${q(
+        moduleSpecifier,
+      )} in compartment ${q(compartment.name)}`;
+    }
+    moduleDescriptor = moduleHook(moduleSpecifier);
+    // Uninitialized module namespaces throw if we attempt to coerce them into
+    // promises.
+    if (!weakmapHas(moduleAliases, moduleDescriptor)) {
+      moduleDescriptor = yield moduleDescriptor;
+    }
   }
-  if (typeof aliasNamespace === 'string') {
+
+  if (typeof moduleDescriptor === 'string') {
     // eslint-disable-next-line @endo/no-polymorphic-call
     assert.fail(
```

Comment:
This presumably means that the type isn’t getting properly threaded, since `fail` and `Fail` return `never`. I’ll change this to `throw assert.error` since throwing at an inevitably throwing call feels weird.

Reply:
It felt weird for me the first 100 times. But our codebase now does that pervasively. In `throw Fail...` I read the `Fail` as "throw an error" and the `throw` as "Let static analysis know that we're throwing an error".